README.md
loresim2
========

Long read simulation.

Source
------

The loresim source code is hosted on github:

	git@github.com:gt1/loresim.git

Compilation of loresim
----------------------

loresim needs libmaus2 [https://github.com/gt1/libmaus2] . When libmaus2
is installed in ${LIBMAUSPREFIX} then biobambam2 can be compiled and
installed in ${HOME}/biobambam2 using

	- autoreconf -i -f
	- ./configure --with-libmaus2=${LIBMAUSPREFIX} \
		--prefix=${HOME}/biobambam2
	- make install

Contained programs
------------------

*fastareformat*: reformats FastA or FastQ so it becomes valid input for the DAZZ_DB tool fasta2DB. This includes changing read names and wrapping long lines.

*longreadsim*: generates a sequence of random DNA reads from a given sequence. For this the input sequence is traversed a number of times and split into pieces which are subsequently either
modified within a certain error range or dropped with a certain rate. The program can switch between a low and high error profile during read generation. Reads are randomly taken from the forward or reverse strand.
The program has the following options

- readlenavg: average read length (default: 15k)
- readlenstddev: standard deviation for read length (default: 3k)
- droprate: probability for dropping a read (default: 0.01)
- numtraversals: number of traversals of the input sequence (default: 1)
- startlowprob: probability of starting a read in low error mode (default: 0.7)
- keeplowstate: probability of staying in low error mode when in low error mode (default: 0.9998)
- keephighstate: probability of staying in high error mode when in high error mode (default: 0.995)
- eratelow: low error mode error rate average
- eratelowstddev: standard deviation of error rate in low error mode
- eratehigh: high error mode error rate average
- eratehighstddev: standard deviation of error rate in high error mode
- placerandom: randomly place reads instead of performing linear traversal (default 0). This produces output data with uneven coverage following a Poisson distribution.

ChangeLog
loresim2 (0.0.17-1) unstable; urgency=medium

  * fix crash

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jun 2018 11:04:30 +0200

loresim2 (0.0.16-1) unstable; urgency=medium

  * avoid front and back deletions in longreadgen
  * do not load Rintv vector in mapcheckstrict (currently not used anyway)

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 05 Jun 2018 09:52:22 +0200

loresim2 (0.0.15-1) unstable; urgency=medium

  * update libmaus2 version
  * add estimatelognormal program
  * update documentation and use log normal length distribution in longreadgen

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 25 Sep 2017 13:39:33 +0200

loresim2 (0.0.14-1) unstable; urgency=medium

  * add minlen parameter in longreadgen

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 23 Jun 2017 11:08:29 +0200

loresim2 (0.0.13-1) unstable; urgency=medium

  * set default drop rate to 0

 -- German Tischler <tischler@mpi-cbg.de>  Tue, 25 Apr 2017 15:29:59 +0200

loresim2 (0.0.12-1) unstable; urgency=medium

  * add nthres parameter in longreadgen

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 10 Nov 2016 14:07:04 +0100

loresim2 (0.0.11-1) unstable; urgency=medium

  * make writing of output files optional in mapcheckstrict 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 22 Aug 2016 14:11:41 +0200

loresim2 (0.0.10-1) unstable; urgency=medium

  * keep only checking of secondary flag for detecting primary alignments

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 28 Jul 2016 16:50:08 +0200

loresim2 (0.0.9-1) unstable; urgency=medium

  * add repeat checking in mapcheckstrict 

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 28 Jul 2016 10:26:52 +0200

loresim2 (0.0.8-1) unstable; urgency=medium

  * print more info in progress report of mapcheckstrict 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 18 Jul 2016 16:51:13 +0200

loresim2 (0.0.7-1) unstable; urgency=medium

  * updates in mapcheckstrict:
  * create primary_no_cross.bam and primary_no_overlap.bam files add 
  * skipa,skipb,unmapped,g_anycross,g_nocross,g_anyoverlap,g_nooverlap,g_primaryanycross,g_primarynocross,g_primaryanyoverlap and g_primarynooverlap in progress lines

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 18 Jul 2016 14:41:55 +0200

loresim2 (0.0.6-1) unstable; urgency=medium

  * bump libmaus2 version to 2.0.217
  * randomly select strand for each read in longreadgen (instead of choosing once per ref seq traversal)
  * implement random read placement as alternative to splitting of ref seq in longreadgen
  * print number of primary alignments on correct sequence/stand in verbose output in mapcheckstrict

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 04 Jul 2016 13:38:33 +0200

loresim2 (0.0.5-1) unstable; urgency=medium

  * print more information in printmatchingkmers
  * write crossing and overlapping alignments to files in mapcheckstrict

 -- German Tischler <tischler@mpi-cbg.de>  Fri, 17 Jun 2016 10:39:18 +0200

loresim2 (0.0.4-1) unstable; urgency=medium

  * cleanup and output missing reads in mapcheckstrict even if not fasta reads file is given

 -- German Tischler <tischler@mpi-cbg.de>  Thu, 16 Jun 2016 13:25:15 +0200

loresim2 (0.0.3-1) unstable; urgency=medium

  * various improvements in mapcheckstrict (self alignment/tandem repeat checks, high error rate intervals)
  * add printmatchingkmers program
  * set default homopolymer insert rate to 0 in longreadgen

 -- German Tischler <tischler@mpi-cbg.de>  Sun, 12 Jun 2016 16:42:25 +0200

loresim2 (0.0.2-1) unstable; urgency=medium

  * add counting of unmapped reads in mapcheckstrict 

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 29 Feb 2016 13:05:50 +0100

loresim2 (0.0.1-1) unstable; urgency=medium

  * Initial packaging release 

 -- German Tischler <tischler@mpi-cbg.de>  Sat, 27 Feb 2016 14:16:40 +0100

loresim2 (0.0.0-1) UNRELEASED; urgency=medium

  * Initial release

 -- German Tischler <tischler@mpi-cbg.de>  Mon, 27 Jan 2016 16:10:29 +0200


src/Intv.hpp
/*
    loresim
    Copyright (C) 2016 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#if ! defined(INTV_HPP)
#define INTV_HPP

#include <libmaus2/math/IntegerInterval.hpp>
#include <libmaus2/util/stringFunctions.hpp>
#include <libmaus2/bambam/BamAlignment.hpp>
#include <sstream>

struct Intv
{
	bool valid;
	libmaus2::math::IntegerInterval<int64_t> intv;
	double erate;
	int64_t numerr;
	int64_t ins;
	int64_t sub;
	int64_t del;

	static bool expect(std::istream & in, std::string const & s)
	{
		uint64_t i = 0;
		bool ok = true;
		while ( ok && i < s.size() && in.peek() != std::istream::traits_type::eof() )
			ok = ok && (s[i++] == in.get());
		return ok;
	}

	static bool getNumber(std::istream & in, int64_t & num)
	{
		bool ok = false;
		num = 0;
		while ( in.peek() != std::istream::traits_type::eof() && isdigit(in.peek()) )
		{
			num *= 10;
			num += in.get()-'0';
			ok = true;
		}

		return ok;
	}

	static bool getUntilTerm(std::istream & in, std::ostream & ostr, int term)
	{
		while ( in.peek() != std::istream::traits_type::eof() && in.peek() != term )
			ostr.put(in.get());

		if ( in.peek() != std::istream::traits_type::eof() )
		{
			int const t = in.get();
			assert ( t == term );
			return true;
		}
		else
		{
			return false;
		}
	}

	static bool getFloat(std::istream & in, double & d, int term)
	{
		std::ostringstream ostr;
		bool ok = getUntilTerm(in,ostr,term);
		std::istringstream istr(ostr.str());
		istr >> d;
		ok = ok && istr;
		return ok;
	}

	bool setup(std::istream & in)
	{
		bool ok = true;

		ok = ok && expect(in,"intv([");
		ok = ok && getNumber(in,intv.from);
		ok = ok && expect(in,",");
		ok = ok && getNumber(in,intv.to);
		if ( ok )
			intv.to -= 1;
		ok = ok && expect(in,"),erate=");
		ok = ok && getFloat(in,erate,',');
		ok = ok && expect(in,"numerr=");
		ok = ok && getNumber(in,numerr);
		ok = ok && expect(in,",ins=");
		ok = ok && getNumber(in,ins);
		ok = ok && expect(in,",sub=");
		ok = ok && getNumber(in,sub);
		ok = ok && expect(in,",del=");
		ok = ok && getNumber(in,del);
		ok = ok && expect(in,")");
		return ok;
	}

	Intv() : valid(false)
	{

	}

	Intv(std::istream & in)
	: valid(false)
	{
		valid = setup(in);
	}

	Intv(std::string const & s)
	: valid(false)
	{
		std::istringstream istr(s);
		valid = setup(istr);
	}

	static std::vector<Intv> parse(std::string const & s)
	{
		std::vector<Intv> Vintv;

		std::deque<std::string> tokens = libmaus2::util::stringFunctions::tokenize<std::string>(s,std::string(";"));

		for ( uint64_t i = 0; i < tokens.size(); ++i )
			if ( tokens[i].size() && tokens[i].size() >= strlen("intv") && tokens[i].substr(0,strlen("intv")) == std::string("intv") )
			{
				Intv intv(tokens[i]);
				if (intv.valid )
					Vintv.push_back(intv);
			}

		for ( uint64_t i = 1; i < Vintv.size(); ++i )
			assert ( Vintv[i-1].intv.to+1 == Vintv[i].intv.from );

		return Vintv;
	}

	static std::vector<Intv> computeRIntv(std::vector<Intv> const & Vintv, libmaus2::bambam::BamAlignment const & a_algn)
	{
		return computeRIntv(Vintv, a_algn.D.get());
	}

	static std::vector<Intv> computeRIntv(std::vector<Intv> const & Vintv, uint8_t const * D)
	{
		assert ( static_cast<int64_t>(libmaus2::bambam::BamAlignmentDecoderBase::getReferenceLength(D)) == Vintv.back().intv.to+1 );

		libmaus2::autoarray::AutoArray<libmaus2::bambam::cigar_operation> cigop;
		uint64_t const numcig = libmaus2::bambam::BamAlignmentDecoderBase::getCigarOperations(D,cigop);
		std::vector < libmaus2::bambam::BamFlagBase::bam_cigar_ops > expcig;
		for ( uint64_t i = 0; i < numcig; ++i )
			for ( int64_t j = 0; j < cigop[i].second; ++j )
				expcig.push_back(static_cast<libmaus2::bambam::BamFlagBase::bam_cigar_ops>(cigop[i].first));
		uint64_t zz = 0;
		while (
			zz < expcig.size() &&
			(expcig[zz] == libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CSOFT_CLIP ||
			 expcig[zz] == libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CHARD_CLIP )
		)
			++zz;

		std::vector<Intv> Rintv;
		uint64_t rlsum = 0;
		for ( uint64_t zi = 0; zi < Vintv.size(); ++zi )
		{
			// length on reference
			uint64_t d = Vintv[zi].intv.to-Vintv[zi].intv.from+1;
			// length on read
			uint64_t rl = 0;

			uint64_t ins = 0, del = 0, match = 0, mismatch = 0;

			while ( d )
			{
				switch ( expcig[zz++] )
				{
					case libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CMATCH:
						--d;
						++rl;
						break;
					case libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CEQUAL:
						match += 1;
						--d;
						++rl;
						break;
					case libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CDIFF:
						mismatch += 1;
						--d;
						++rl;
						break;
					case libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CINS:
						ins += 1;
						++rl;
						break;
					case libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CDEL:
						del += 1;
						--d;
						break;
					default:
						break;
				}
			}

			while ( zz < expcig.size() && expcig[zz] == libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CINS )
			{
				++zz;
				++rl;
			}

			Intv rintv;
			rintv.valid = true;
			rintv.intv.from = rlsum;
			rintv.intv.to = rintv.intv.from + rl - 1;
			rintv.ins = ins;
			rintv.del = del;
			rintv.sub = mismatch;
			rintv.numerr = ins+del+mismatch;
			rintv.erate = (ins+del+mismatch) / static_cast<double>(ins+del+mismatch+match);
			Rintv.push_back(rintv);

			rlsum += rl;
		}

		assert (
			static_cast<int64_t>(
				rlsum+
				libmaus2::bambam::BamAlignmentDecoderBase::getFrontSoftClipping(D) +
				libmaus2::bambam::BamAlignmentDecoderBase::getBackSoftClipping(D)
			)
			==
			libmaus2::bambam::BamAlignmentDecoderBase::getLseq(D)
		);

		return Rintv;
	}
};

inline std::ostream & operator<<(std::ostream & out, Intv const & intv)
{
	return out << "intv([" << intv.intv.from << "," << intv.intv.to+1 << "),erate=" << intv.erate << ",numerr=" << intv.numerr << ",ins=" << intv.ins << ",sub=" << intv.sub << ",del=" << intv.del << ")";
}
#endif


src/fastareformat.cpp
/*
    loresim
    Copyright (C) 2015 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/fastx/StreamFastAReader.hpp>
#include <libmaus2/fastx/StreamFastQReader.hpp>
#include <libmaus2/aio/PosixFdInputStream.hpp>
#include <libmaus2/fastx/SpaceTable.hpp>
#include <libmaus2/random/Random.hpp>

template<typename reader_type>
void fastareformat(libmaus2::util::ArgInfo const & arginfo, reader_type & reader)
{
	typename reader_type::pattern_type pattern;

	std::string const prolog = arginfo.getUnparsedValue("prolog","L");
	uint64_t readid = arginfo.getValueUnsignedNumeric<uint64_t>("readidbase",0);
	uint64_t const cols = arginfo.getValueUnsignedNumeric<uint64_t>("cols",80);
	bool const randomN = arginfo.getValueUnsignedNumeric<uint64_t>("randomN",true);
	libmaus2::fastx::SpaceTable const ST;
	libmaus2::random::Random::setup();
	bool replaced = false;
	bool replaceprinted = false;

	while ( reader.getNextPatternUnlocked(pattern) )
	{
		std::string s = pattern.spattern;

		if ( randomN )
			for ( uint64_t i = 0; i < s.size(); ++i )
				if ( libmaus2::fastx::mapChar(s[i]) >= 4 )
				{
					s[i] = libmaus2::fastx::remapChar(libmaus2::random::Random::rand8() & 3);
					replaced = true;
				}
		char const * c = s.c_str();

		if ( replaced && ! replaceprinted )
		{
			std::cerr << "[V] warning, replacing non ACGT symbols with random bases" << std::endl;
			replaceprinted = true;
		}

		std::cout << '>' << prolog << '/' << (readid++) << '/' << 0 << '_' << s.size() << " RQ=0.851 " << pattern.sid << "\n";
		uint64_t low = 0;

		while ( low < s.size() )
		{
			uint64_t const high = std::min(low+cols,static_cast<uint64_t>(s.size()));

			std::cout.write(c+low,high-low);
			std::cout.put('\n');

			low = high;
		}
	}
}

int main(int argc, char * argv[])
{
	try
	{
		libmaus2::util::ArgInfo const arginfo(argc,argv);

		libmaus2::aio::PosixFdInputStream PFIS(STDIN_FILENO,64*1024,1024);

		int const c = PFIS.peek();

		if ( c == '>' || c == std::istream::traits_type::eof() )
		{
			libmaus2::fastx::StreamFastAReaderWrapper in(PFIS);
			fastareformat(arginfo,in);
		}
		else if ( c == '@' )
		{
			libmaus2::fastx::StreamFastQReaderWrapper in(PFIS);
			fastareformat(arginfo,in);
		}
		else
		{
			libmaus2::exception::LibMausException lme;
			lme.getStream() << "Unknown input file format, first character " << static_cast<char>(c) << std::endl;
			lme.finish();
			throw lme;
		}
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}


src/estimatelognormal.cpp
/*
    loresim2
    Copyright (C) 2017 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/random/LogNormalRandom.hpp>
#include <libmaus2/fastx/StreamFastAReader.hpp>
#include <iostream>

int main()
{
	try
	{
		std::vector<uint64_t> V;

		libmaus2::fastx::StreamFastAReaderWrapper SFAR(std::cin);
		libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;
		uint64_t tl = 0;

		while ( SFAR.getNextPatternUnlocked(pattern) )
		{
			uint64_t const l = pattern.spattern.size();

			if ( l )
			{
				while ( ! (l < V.size()) )
					V.push_back(0);

				assert ( l < V.size() );

				V[l]++;
				tl += l;
			}
		}

		std::pair<double,double> const P = libmaus2::random::LogNormalRandom::computeAverageAndSigma(V);
		std::pair<double,double> const PV = libmaus2::random::LogNormalRandom::computeParameters(P.first,P.second);
		std::cout << PV.first << "\t" << PV.second << "\t" << P.first << "\t" << P.second << std::endl;
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}


src/longreadgen.cpp
/*
    loresim
    Copyright (C) 2015 German Tischler

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <libmaus2/bambam/BamHeader.hpp>
#include <libmaus2/bambam/BamBlockWriterBaseFactory.hpp>
#include <libmaus2/fastx/StreamFastAReader.hpp>
#include <libmaus2/random/DNABaseNoiseSpiker.hpp>
#include <libmaus2/util/ArgInfo.hpp>
#include <libmaus2/fastx/FastAIndex.hpp>
#include <libmaus2/fastx/FastAIndexGenerator.hpp>
#include <libmaus2/random/LogNormalRandom.hpp>
#include "Intv.hpp"

std::string getRandom(uint64_t const randlen)
{
	std::ostringstream o;

	for ( uint64_t i = 0; i < randlen; ++i )
	{
        	switch ( libmaus2::random::Random::rand8() % 4 )
                {
                	case 0: o.put('A'); break;
                	case 1: o.put('C'); break;
                	case 2: o.put('G'); break;
                	case 3: o.put('T'); break;
		}
	}

	return o.str();
}

int main(int argc, char ** argv)
{
	try
	{
		libmaus2::util::ArgInfo const arginfo(argc,argv);
		uint64_t randseed = arginfo.getValueUnsignedNumeric<uint64_t>("randomseed",time(0));
		libmaus2::random::Random::setup(randseed);

		libmaus2::bambam::BamSeqEncodeTable const seqenc;
		::libmaus2::fastx::UCharBuffer UB;

		// substitution error fraction
		double substrate = arginfo.getValue<double>("substrate",.2);
		// deletion error fraction
		double delrate = arginfo.getValue<double>("delrate",.3);
		// insertion error fraction
		double insrate = arginfo.getValue<double>("insrate",.5);
		// insertion homopolymer error fraction
		double inshomopolrate = arginfo.getValue<double>("inshomopolrate",.0);

		// low error rate
		double eratelow        = arginfo.getValue<double>("eratelow",.15);
		// high error rate
		double eratehigh       = arginfo.getValue<double>("eratehigh",.25);
		// std dev for low error rate
		double eratelowstddev  = arginfo.getValue<double>("eratelowstddev",0.03);
		// std dev for high error rate
		double eratehighstddev = arginfo.getValue<double>("eratehighstddev",0.04);

		// 8581.35	6953.55

		// drop rate
		double const droprate = arginfo.getValue<double>("droprate",0.00);
		// number of traversals per input sequence
		uint64_t const numtraversals = arginfo.getValueUnsignedNumeric<uint64_t>("numtraversals",1);
		// average read length
		double const readlenavg = arginfo.getValue<double>("readlenavg", 8500);
		// read length std dev
		double const readlenstddev = arginfo.getValue<double>("readlenstddev", 6953);
		// probability to stay in low error rate mode
		double const keeplowstate = arginfo.getValue<double>("keeplowstate", 0.9998);
		// probability to stay in high error rate mode
		double const keephighstate = arginfo.getValue<double>("keephighstate", 0.995);
		// probability to start in low error rate mode
		double const startlowprob = arginfo.getValue<double>("startlowprob", 0.7);
		// number of random bases appended at front and back
		uint64_t const randlen = arginfo.getValueUnsignedNumeric<uint64_t>("randlen",0);
		bool const placerandom = arginfo.getValueUnsignedNumeric<uint64_t>("placerandom",0);
		bool const nthres = arginfo.getValueUnsignedNumeric<uint64_t>("nthres",0);

		uint64_t const minlen = arginfo.getValueUnsignedNumeric<uint64_t>("minlen",0);

		std::pair<double,double> const lognormalP = libmaus2::random::LogNormalRandom::computeParameters(readlenavg,readlenstddev);
		// std::floor(libmaus2::random::LogNormalRandom::random(lognormalP.second, lognormalP.first) + 0.5);

		// noise spiker object
		libmaus2::random::DNABaseNoiseSpiker DBNS(substrate, delrate, insrate, inshomopolrate, eratelow, eratehigh, eratelowstddev, eratehighstddev, keeplowstate, keephighstate, startlowprob);

		if ( arginfo.restargs.size() < 2 )
		{
			std::cerr << "usage: " << arginfo.progname << " <text.fa> <out.fasta> >out.bam" << std::endl;
			return EXIT_FAILURE;
		}

		// input reference name (FastA format)
		std::string const reffn = arginfo.restargs.at(0);
		// output fasta file name
		std::string const fafn = arginfo.restargs.at(1);

		// construct name of FAI file
		std::string const fainame = reffn + ".fai";
		libmaus2::fastx::FastAIndexGenerator::generate(reffn,fainame,1);
		// load FAI index for reference FastA
		libmaus2::fastx::FastAIndex::unique_ptr_type PFAI(libmaus2::fastx::FastAIndex::load(fainame));
		// open output fasta file
		libmaus2::aio::OutputStreamInstance faOSI(fafn);

		// create SAM header
		std::ostringstream samheaderstr;
		samheaderstr << "@HD\tVN:1.5\tSO:unknown\n";
		for ( uint64_t i = 0; i < PFAI->size(); ++i )
		{
			libmaus2::fastx::FastAIndexEntry const & entry = (*PFAI)[i];
			samheaderstr << "@SQ\tSN:" << entry.name << "\tLN:" << entry.length << "\n";
		}

		libmaus2::bambam::BamHeader header(samheaderstr.str());

		// bam writer
		libmaus2::bambam::BamBlockWriterBase::unique_ptr_type writer(
			libmaus2::bambam::BamBlockWriterBaseFactory::construct(header, arginfo)
		);

		// id of run
		uint64_t const runid = 0;
		// current well id
		uint64_t wellid = 0;

		libmaus2::aio::InputStreamInstance ISI(reffn);
		libmaus2::fastx::StreamFastAReaderWrapper SFAR(ISI);
		libmaus2::fastx::StreamFastAReaderWrapper::pattern_type pattern;

		for ( uint64_t seqid = 0; SFAR.getNextPatternUnlocked(pattern); ++seqid )
		{
			std::cerr << "[V] " << pattern.sid << std::endl;

			std::ostringstream seqstr;
			seqstr << pattern.spattern;

			std::string const seq = seqstr.str();

			for ( uint64_t trav = 0; trav < numtraversals; ++trav )
			{
				std::cerr << "\t[V] traversal " << trav << std::endl;

				uint64_t pp = 0;
				typedef std::pair<uint64_t,uint64_t> upair;
				std::vector<upair> poslenvec;

				if ( placerandom )
					while ( pp < seq.size() )
					{
						int64_t len = -1;
						while ( len <= 0 )
							len = std::floor(libmaus2::random::LogNormalRandom::random(lognormalP.second, lognormalP.first) + 0.5);
							// libmaus2::random::GaussianRandom::random(readlenstddev,readlenavg);
						len = std::min(len,static_cast<int64_t>(seq.size()));
						uint64_t p = libmaus2::random::Random::rand64() % (seq.size()-len+1);
						poslenvec.push_back(upair(p,len));
						pp += len;
					}
				else
					while ( pp < seq.size() )
					{
						int64_t len = -1;
						while ( len <= 0 )
							len = std::floor(libmaus2::random::LogNormalRandom::random(lognormalP.second, lognormalP.first) + 0.5);
							// libmaus2::random::GaussianRandom::random(readlenstddev,readlenavg);

						len = std::min(len,static_cast<int64_t>(seq.size()-pp));

						poslenvec.push_back(upair(pp,len));

						pp += len;
					}

				for ( uint64_t z = 0; z < poslenvec.size(); ++z )
				{
					if ( libmaus2::random::UniformUnitRandom::uniformUnitRandom() < droprate )
						continue;
					uint64_t const len = poslenvec[z].second;
					if ( len < minlen )
						continue;

					uint64_t const pos = poslenvec[z].first;
					bool const strand = libmaus2::random::Random::rand8() & 1;
					bool rc = !strand;
					std::string sub = seq.substr(pos,len);

					uint64_t ncnt = 0;
					for ( uint64_t i = 0; i < sub.size(); ++i )
						if ( sub[i] == 'N' )
							ncnt++;
					if ( ncnt > nthres )
						continue;

					libmaus2::random::DNABaseNoiseSpiker::ErrorStats E;
					std::pair<std::string,std::string> const P = DBNS.modifyAndComment(sub,&E);

					// std::cerr << P.second << std::endl;

					std::string errintv = P.second;
					if ( errintv.find("ERRINTV=[[") != std::string::npos )
					{
						errintv = errintv.substr(errintv.find("ERRINTV=[[" )+ strlen("ERRINTV=[["));
						if ( errintv.find("]]") != std::string::npos )
						{
							errintv = errintv.substr(0,errintv.find("]]"));
							#if 0
							std::deque<std::string> Vintv(libmaus2::util::stringFunctions::tokenize<std::string>(errintv,";"));
							for ( uint64_t i = 0; i < Vintv.size(); ++i )
								std::cerr << Vintv[i] << std::endl;
							#endif
						}
						else
						{
							errintv = std::string();
						}
					}
					else
					{
						errintv = std::string();
					}

					std::string mod = P.first;

					std::string const r = getRandom(randlen) + mod + getRandom(randlen);

					std::string cig = P.second;
					if ( cig.find("CIGAR=[") != std::string::npos )
						cig = cig.substr(cig.find("CIGAR=[")+strlen("CIGAR=["));
					if ( cig.find("]") != std::string::npos )
						cig = cig.substr(0,cig.find("]"));

					std::vector<libmaus2::bambam::cigar_operation> Vcigop = libmaus2::bambam::CigarStringParser::parseCigarString(cig);
					libmaus2::lcs::AlignmentTraceContainer ATC;
					libmaus2::bambam::CigarStringParser::cigarToTrace(Vcigop.begin(),Vcigop.end(),ATC);
					uint64_t frontdel = 0;
					uint64_t backdel = 0;
					while ( ATC.ta != ATC.te && ATC.ta[0] == libmaus2::lcs::AlignmentTraceContainer::STEP_DEL )
					{
						++frontdel;
						++ATC.ta;
					}
					while ( ATC.ta != ATC.te && ATC.te[-1] == libmaus2::lcs::AlignmentTraceContainer::STEP_DEL )
					{
						++backdel;
						--ATC.te;
					}

					libmaus2::autoarray::AutoArray< std::pair<libmaus2::lcs::AlignmentTraceContainer::step_type,uint64_t> > Aopblocks;
	                                libmaus2::autoarray::AutoArray<libmaus2::bambam::cigar_operation> Aop;
	                                uint64_t const ncig = libmaus2::bambam::CigarStringParser::traceToCigar(ATC,Aopblocks,Aop,0,0,0,0);
	                                std::string const upcig = libmaus2::bambam::CigarStringParser::cigarBlocksToString(Aop.begin(),ncig);

	                                #if 0
					uint64_t delshift = 0;
					for (
						uint64_t ind = 0;
						ind < Vcigop.size() &&
						Vcigop[ind].first != libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CMATCH &&
						Vcigop[ind].first != libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CEQUAL &&
						Vcigop[ind].first != libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CDIFF
						; ++ind
					)
					{
						if ( Vcigop[ind].first == libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_CDEL )
							delshift += Vcigop[ind].second;
					}
					#endif

					std::ostringstream padcigstr;
					if ( randlen )
						padcigstr << randlen << "S";
					padcigstr << upcig;
					if ( randlen )
						padcigstr << randlen << "S";
					cig = padcigstr.str();

					std::ostringstream namestr;
					std::string const rfasta = strand ? r : libmaus2::fastx::reverseComplementUnmapped(r);
					namestr << 'L' << runid << '/' << (wellid++) << '/' << 0 << '_' << rfasta.size();
					std::string const name = namestr.str();

					UB.reset();
					std::vector<char> Vqual(r.size(),'H');

					try
					{
						libmaus2::bambam::BamAlignmentEncoderBase::encodeAlignment
						(
							UB,
							seqenc,
							name,
							seqid,
							pos + frontdel,
							255,
							rc ? libmaus2::bambam::BamFlagBase::LIBMAUS2_BAMBAM_FREVERSE : 0,
							cig,
							-1,
							-1,
							0,
							r,
							std::string(Vqual.begin(),Vqual.end()),
							33,true
						);
					}
					catch(std::exception const & lme)
					{
						std::cerr << lme.what() << std::endl;
						continue;
					}

					::libmaus2::bambam::MdStringComputationContext mdnmcontext;
					libmaus2::bambam::BamAlignmentDecoderBase::calculateMd(UB.buffer,UB.length,mdnmcontext,seq.begin()+pos /* itref */);

					libmaus2::bambam::BamAlignmentEncoderBase::putAuxString(UB,"MD",mdnmcontext.md.get());
					libmaus2::bambam::BamAlignmentEncoderBase::putAuxNumber(UB,"NM",'i',mdnmcontext.nm);

					if ( errintv.size() )
					{
						libmaus2::bambam::BamAlignmentEncoderBase::putAuxString(
							UB,
							"er",
							errintv.c_str()
						);

						#if 0
						std::vector<Intv> Vintv = Intv::parse(errintv);
						std::vector<Intv> Rintv = Intv::computeRIntv(Vintv,UB.buffer);

						std::ostringstream ostr;
						for ( uint64_t i = 0; i < Rintv.size(); ++i )
							ostr << Rintv[i] << ";";

						libmaus2::bambam::BamAlignmentEncoderBase::putAuxString(
							UB,
							"ee",
							ostr.str().c_str()
						);
						#endif
					}

					libmaus2::bambam::BamAlignmentEncoderBase::putAuxString(UB,"oc",cig.c_str());

					faOSI << '>' << name << " RQ=0.851" << " POS=[" << pos << "] REFID=[" << seqid << "] RC=[" << rc << "]\n";
					for ( uint64_t i = 0; i < rfasta.size(); )
					{
						uint64_t toprint = std::min(static_cast<int>(rfasta.size()-i),static_cast<int>(80));
						faOSI.write(rfasta.c_str()+i,toprint);
						i += toprint;
						faOSI.put('\n');
					}
					writer->writeBamBlock(UB.buffer,UB.length);
				}
			}
		}
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
